### PR TITLE
Task for bank with TDD

### PR DESCRIPTION
**Pull request by Stanislava88**

### CODE REVIEW

**Comment by freekman** [comment]
You should add a description for your PL not only the title 
Ex :
```js
Here i have solved the task using TDD ..... 
```

File: .idea/modules.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (freekman) [inline .idea/modules.xml]: exclude ./idea in the .gitignore

File: src/com/clouway/bank/Account.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package com.clouway.bank;
+
+/**
+ * @author Stanislava Kaukova(sisiivanovva@gmail.com))
```
Comment (freekman) [inline src/com/clouway/bank/Account.java]: here you have an extra ``` ) ```

File: src/com/clouway/bank/Account.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ */
+public class Account {
+    public final double limit;
+    private final String userName;
+    private double cash;
+
+
```
Comment (freekman) [inline src/com/clouway/bank/Account.java]: format code

File: src/com/clouway/bank/Bank.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public void add(int id, Account account) {
+        volt.put(id, account);
+    }
+
+    public Account findById(int id) {
+        if (volt.containsKey(id)) {
+            return volt.get(id);
+        } else throw new AccountNotFoundException("This user is not available");
```
Comment (freekman) [inline src/com/clouway/bank/Bank.java]: here you are throwing AccountNotFoundException and say ```User  is not available``` eather change the Exception to UserNotFoudException or change the message

File: src/com/clouway/bank/Bank.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ */
+public class Bank {
+    private HashMap<Integer, Account> volt;
+
+    public Bank(HashMap<Integer, Account> volt) {
+
+        this.volt = volt;
+    }
+
+    public void add(int id, Account account) {
```
Comment (freekman) [inline src/com/clouway/bank/Bank.java]: maybe ```createAccount``` would be bether here

File: tests/com/clouway/bank/BankTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.HashMap;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Stanislava Kaukova(sisiivanovva@gmail.com))
+ */
+public class BankTest {
+    Bank bank = new Bank(new HashMap<Integer, Account>());
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: make Bank private

File: tests/com/clouway/bank/BankTest.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Test
+    public void happyPath() throws Exception {
+        bank.add(12345, new Account("Ivan", 200, 1000));
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: you could extract Account and reuse it 
```js
 +        Account account= new Account("Ivan", 200, 1000)

 +        bank.add(12345, account);
 +        Account result = bank.findById(12345);
 +
 +        assertThat(result, is(account)));
```

File: tests/com/clouway/bank/BankTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Test(expected = AccountNotFoundException.class)
+    public void detectedIsMissingAccount() throws Exception {
+        bank.findById(12);
+    }
+
+    @Test
+    public void withDraw() throws Exception {
+        bank.add(101, new Account("Lilia", 500, 1000));
+        bank.withDraw(101, 200);
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: i think ```bank.withDraw(new Id(101), 200);``` would be more readable than ```bank.withDraw(101, 200);```

File: tests/com/clouway/bank/BankTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        bank.deposit(100, 200);
+        Account result = bank.findById(100);
+
+        assertThat(result, is(new Account("Maria", 500, 1000)));
+    }
+
+    @Test(expected = InsufficientCashException.class)
+    public void insufficientCash() throws Exception {
+        bank.add(102, new Account("Ivan", 20, 1000));
+        bank.withDraw(102, 100);
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: what will happen if we withDraw from an Undefined user

File: tests/com/clouway/bank/BankTest.java (hunk 7 above)
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: I i add an user with the same Id will i get an error

**Comment by Stanislava88** [comment]
Done

**Comment by Stanislava88** [comment]
done

File: .gitignore [hunk 10]
```diff
@@ -1,4 +1,7 @@
 *.class
+.idea/workspace.xml
```
Comment (freekman) [inline .gitignore]: just .idea

File: tests/com/clouway/bank/BankTest.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        bank.findById(12);
+    }
+
+    @Test
+    public void withDraw() throws Exception {
+        int id = 101;
+
+        bank.createAccount(id, new Account("Lilia", 500, 1000));
+        bank.withDraw(id, 200);
+        Account result = bank.findById(101);
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: replace ```101``` with ```int id```

**Comment by Stanislava88** [comment]
done

File: tests/com/clouway/bank/BankTest.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        assertThat(result, is(new Account("Lilia", 300, 1000)));
+    }
+
+    @Test
+    public void deposit() throws Exception {
+        int id = 100;
+
+        bank.createAccount(id, new Account("Maria", 300, 1000));
+        bank.deposit(id, 200);
+        Account result = bank.findById(100);
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: .

File: tests/com/clouway/bank/BankTest.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+        assertThat(result, is(new Account("Maria", 500, 1000)));
+    }
+
+    @Test(expected = InsufficientCashException.class)
+    public void insufficientCash() throws Exception {
+        int id = 102;
+
+        bank.createAccount(id, new Account("Ivan", 20, 1000));
+        bank.withDraw(102, 100);
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: .

File: tests/com/clouway/bank/BankTest.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    @Test(expected = AccountNotFoundException.class)
+    public void withDrawToUndefineAccount() throws Exception {
+        int id = 102;
+        bank.withDraw(id, 100);
+    }
+
+    @Test(expected = ExceedLimitException.class)
+    public void exceedLimit() throws Exception {
+        int id = 103;
+        bank.createAccount(103, new Account("Krasimir", 20, 1000));
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: .

**Comment by Stanislava88** [comment]
fixed

File: tests/com/clouway/bank/BankTest.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+            bank.findById(id);
+
+            fail("Should be thrown exception");
+        } catch (AccountNotFoundException ex) {
+            assertThat(ex.getMessage(), is(equalTo("This account is removed")));
+        }
+    }
+
+    @Test
+    public void removeUnregisterAccount() throws Exception {
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: - i cant understand what  are you testing here
- the test is hard which is a  clear sign something is wrong

To test remove : 
 - remove a account 
 - remove an undefined account 

```js 
bank.save(account)
bank.remove(AccountId)

bank.Find(AccountId)
```

File: tests/com/clouway/bank/BankTest.java [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        bank.createAccount(101, new Account("Lilia", 500, 1000));
+        bank.createAccount(102, new Account("Petia", 50, 1000));
+        bank.createAccount(103, new Account("Ivo", 100, 1000));
+        ArrayList<Integer> result = bank.sortByCash();
+
+        assertThat(result, is(Arrays.asList(102, 103, 100, 101)));
+    }
+
+    @Test
+    public void expectedExceptionAfterRemove() throws Exception {
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: .

**Comment by Stanislava88** [comment]
made some changes

**Comment by Stanislava88** [comment]
Changed method remove

File: tests/com/clouway/bank/BankTest.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+        assertThat(result, is(Arrays.asList(102, 103, 100, 101)));
+    }
+
+    @Test(expected = AccountNotFoundException.class)
+    public void remove() throws Exception {
+        int accountId = 104;
+
+        Account account = new Account("Maria", 300, 1000);
+        bank.save(accountId, account);
```
Comment (freekman) [inline tests/com/clouway/bank/BankTest.java]: last comment rename save,find and remove 
```js
bank.createAccount
bank.removeAccount
bank.findAccountById
```

**Comment by freekman** [comment]
LGTM

**Comment by Stanislava88** [comment]
Renamed

**Pull request merged**